### ISSUE TITLE
Install on Centos8

### ISSUE THREAD

**Issue by davama** [ltb-project/ppm#25]
Hello,

Attempting to install on centos8. I've installed this before but on centos7.

As far as centos7, these are the packages I need for a successful installation:
> yum install yum-utils gcc wget libdb-devel unzip cracklib-devel openldap-devel

For centos8 , the only difference is symas-openldap-devel
```
> yum install yum-utils gcc wget libdb-devel unzip cracklib-devel symas-openldap-devel
Last metadata expiration check: 0:10:09 ago on Tue 30 Jun 2020 11:05:00 AM EDT.
Package yum-utils-4.0.12-3.el8.noarch is already installed.
Package gcc-8.3.1-5.el8.0.2.x86_64 is already installed.
Package wget-1.19.5-8.el8_1.1.x86_64 is already installed.
Package libdb-devel-5.3.28-37.el8.x86_64 is already installed.
Package unzip-6.0-43.el8.x86_64 is already installed.
Package cracklib-devel-2.9.6-15.el8.x86_64 is already installed.
Package symas-openldap-devel-2.4.50-1.el8.x86_64 is already installed.
Dependencies resolved.
Nothing to do.
Complete!
```

Perhaps i am missing some dependency...
**My output**

```
> git clone https://github.com/ltb-project/ppm.git
Cloning into 'ppm'...
remote: Enumerating objects: 27, done.
remote: Counting objects: 100% (27/27), done.
remote: Compressing objects: 100% (20/20), done.
remote: Total 154 (delta 12), reused 21 (delta 7), pack-reused 127
Receiving objects: 100% (154/154), 58.83 KiB | 753.00 KiB/s, done.
Resolving deltas: 100% (83/83), done.

> cd ppm/

> make CONFIG=/etc/openldap/ppm.conf OLDAP_SOURCES=../../..
gcc -g -O2 -Wall -fpic -DCONFIG_FILE="\"/etc/openldap/ppm.conf\"" -DDEBUG -c -I../../../include -I../../../servers/slapd -DCRACKLIB ppm.c
In function ‘strcpy_safe.part.1’,
    inlined from ‘strcpy_safe’ at ppm.c:47:1:
ppm.c:58:9: warning: ‘strncpy’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
         strncpy(dest, src, n);
         ^~~~~~~~~~~~~~~~~~~~~
ppm.c: In function ‘strcpy_safe’:
ppm.c:55:26: note: length computed here
         int length_src = strlen(src);
                          ^~~~~~~~~~~
gcc -I../../../include -I../../../servers/slapd -shared -o ppm.so ppm.o -lcrack
gcc -g -I../../../include -I../../../servers/slapd -L../../../libraries/liblber/.libs -L../../../libraries/libldap_r/.libs -Wl,-rpath=. -o ppm_test ppm_test.c ppm.so -lldap_r -llber -lcrack
/usr/bin/ld: cannot find -lldap_r
/usr/bin/ld: cannot find -llber
collect2: error: ld returned 1 exit status
make: *** [Makefile:45: ppm_test] Error 1


> make install CONFIG=/etc/openldap/ppm.conf LIBDIR=/usr/lib64/openldap
gcc -I../../../include -I../../../servers/slapd -shared -o ppm.so ppm.o -lcrack
cp -f ppm.so /usr/lib64/openldap
cp -f ppm_test /usr/lib64/openldap
cp: cannot stat 'ppm_test': No such file or directory
make: *** [Makefile:55: install] Error 1
```

Any input is much appreciated

Thanks!

**Comment by coudot** [comment]
See LTB OpenLDAP RPM spec file: https://github.com/ltb-project/openldap-rpm/blob/master/SPECS/openldap-ltb.spec#L262

**Comment by davama** [comment]
Thank you for the reply!

Hope im doing it right now.
I get more output  but still no success :(

```
> make CONFIG=/etc/openldap/ppm.conf OLDAP_SOURCES=../../.. "CRACK_INC=" "CRACK_LIB="
gcc -g -O2 -Wall -fpic -DCONFIG_FILE="\"/etc/openldap/ppm.conf\"" -DDEBUG -c -I../../../include -I../../../servers/slapd  ppm.c
ppm.c: In function ‘check_password’:
ppm.c:354:11: warning: unused variable ‘j’ [-Wunused-variable]
     int i,j;
           ^
ppm.c:346:11: warning: unused variable ‘res’ [-Wunused-variable]
     char* res;
           ^~~
ppm.c:345:11: warning: unused variable ‘fd’ [-Wunused-variable]
     FILE* fd;
           ^~
ppm.c:344:17: warning: unused variable ‘cracklibExt’ [-Wunused-variable]
     char const* cracklibExt[] = { ".hwm", ".pwd", ".pwi" };
                 ^~~~~~~~~~~
ppm.c:343:10: warning: unused variable ‘cracklibDictFiles’ [-Wunused-variable]
     char cracklibDictFiles[3][(VALUE_MAX_LEN+5)];
          ^~~~~~~~~~~~~~~~~
ppm.c:341:9: warning: variable ‘useCracklib’ set but not used [-Wunused-but-set-variable]
     int useCracklib;
         ^~~~~~~~~~~
In function ‘strcpy_safe.part.1’,
    inlined from ‘strcpy_safe’ at ppm.c:47:1:
ppm.c:58:9: warning: ‘strncpy’ specified bound depends on the length of the source argument [-Wstringop-overflow=]
         strncpy(dest, src, n);
         ^~~~~~~~~~~~~~~~~~~~~
ppm.c: In function ‘strcpy_safe’:
ppm.c:55:26: note: length computed here
         int length_src = strlen(src);
                          ^~~~~~~~~~~
gcc -I../../../include -I../../../servers/slapd -shared -o ppm.so ppm.o 
gcc -g -I../../../include -I../../../servers/slapd -L../../../libraries/liblber/.libs -L../../../libraries/libldap_r/.libs -Wl,-rpath=. -o ppm_test ppm_test.c ppm.so -lldap_r -llber 
/usr/bin/ld: cannot find -lldap_r
/usr/bin/ld: cannot find -llber
collect2: error: ld returned 1 exit status
make: *** [Makefile:45: ppm_test] Error 1

> make install CONFIG=/etc/openldap/ppm.conf LIBDIR=/usr/lib64/openldap
gcc -I../../../include -I../../../servers/slapd -shared -o ppm.so ppm.o -lcrack
cp -f ppm.so /usr/lib64/openldap
cp -f ppm_test /usr/lib64/openldap
cp: cannot stat 'ppm_test': No such file or directory
make: *** [Makefile:55: install] Error 1
```

**Comment by davidcoutadeur** [comment]
Hi Dave,


|> /usr/bin/ld: cannot find -lldap_r > /usr/bin/ld: cannot find -llber|


You need to copy ppm source code into the openldap source code, or else 
adapt the gcc command:

|-I../../../include -I../../../servers/slapd|

to point to openldap source code.


David


Le 30/06/2020 à 20:12, Dave a écrit :

**Comment by davama** [comment]
Hello @davidcoutadeur 

Thank you for  replying!

Yes, i have been doing that.... please see all my steps. (the output below is condensed, if want to see it all, please let me know)

1. Download openldap source
```
[root@localhost ~]# yumdownloader --source symas-openldap
symas-openldap-2.4.50-1.el8.src.rpm                                                                                                                                                                        4.6 MB/s | 5.5 MB     00:01    
```
2. unpack the source file
```
[root@localhost ~]# mkdir openldapsrc && cd openldapsrc

[root@localhost openldapsrc]# rpm2cpio ../symas-openldap-2.4.50-1.el8.src.rpm | cpio -idvm
ITS8650.patch
...
...
slapd.service
slapd.tmpfiles
11402 blocks
```
3. get into the openldap source folder and compile
```
[root@localhost openldapsrc]# tar xf openldap-2.4.50.tgz 

[root@localhost openldapsrc]# cd openldap-2.4.50

[root@localhost openldap-2.4.50]# ./configure 
Configuring OpenLDAP 2.4.50-Release ...
checking build system type... x86_64-pc-linux-gnu
...
...
    Add syncprov ...
Please run "make depend" to build dependencies


[root@localhost openldap-2.4.50]# make depend
Making depend in /root/openldapsrc/openldap-2.4.50
  Entering subdirectory include
...
...
make[1]: Leaving directory '/root/openldapsrc/openldap-2.4.50/doc'
```
4. attempting to install ppm. as the INSTALL file says to move into slapd-modules
``` 
[root@localhost openldap-2.4.50]# cd contrib/slapd-modules/

[root@localhost slapd-modules]# git clone https://github.com/ltb-project/ppm.git
Cloning into 'ppm'...
...
...
Resolving deltas: 100% (83/83), done.

[root@localhost slapd-modules]# cd ppm/

[root@localhost ppm]# make CONFIG=/etc/openldap/ppm.conf OLDAP_SOURCES=../../.. "CRACK_INC=" "CRACK_LIB="
gcc -g -O2 -Wall -fpic -DCONFIG_FILE="\"/etc/openldap/ppm.conf\"" -DDEBUG -c -I../../../include -I../../../servers/slapd  ppm.c
ppm.c: In function ‘check_password’:
...
...
gcc -I../../../include -I../../../servers/slapd -shared -o ppm.so ppm.o 
gcc -g -I../../../include -I../../../servers/slapd -L../../../libraries/liblber/.libs -L../../../libraries/libldap_r/.libs -Wl,-rpath=. -o ppm_test ppm_test.c ppm.so -lldap_r -llber 
/usr/bin/ld: cannot find -lldap_r
/usr/bin/ld: cannot find -llber
collect2: error: ld returned 1 exit status
make: *** [Makefile:45: ppm_test] Error 1


[root@localhost ppm]# make install CONFIG=/etc/openldap/ppm.conf LIBDIR=/usr/lib64/openldap
gcc -I../../../include -I../../../servers/slapd -shared -o ppm.so ppm.o -lcrack
cp -f ppm.so /usr/lib64/openldap
cp -f ppm_test /usr/lib64/openldap
cp: cannot stat 'ppm_test': No such file or directory
make: *** [Makefile:55: install] Error 1
```

**Comment by davama** [comment]
Sorry forgot to add.

```
[root@localhost ppm]# ls ../../../include/
ac               lber.h        lber_types.hin  ldap_config.hin  ldap_features.hin  ldap_log.h         ldap_pvt_uc.h  ldap_schema.h  lutil.h       lutil_lockf.h  lutil_sha1.h  openldap.h    rewrite.h       stamp-h2
avl.h            lber_pvt.h    ldap_cdefs.h    ldap_defaults.h  ldap.h             ldap_pvt.h         ldap_queue.h   ldap_utf8.h    lutil_hash.h  lutil_md5.h    Makefile      portable.h    slapi-plugin.h  stamp-h3
getopt-compat.h  lber_types.h  ldap_config.h   ldap_features.h  ldap_int_thread.h  ldap_pvt_thread.h  ldap_rq.h      ldif.h         lutil_ldap.h  lutil_meter.h  Makefile.in   portable.hin  stamp-h1        sysexits-compat.h
[root@localhost ppm]# ls ../../../servers/slapd/
abandon.c   alock.h      backends.c  back-monitor  back-shell   compare.c     cr.c       extended.c     ldapsync.c       modify.c  oc.c           proto-slap.h  schema.c        sets.h          slapcommon.h  slappasswd.c  syntax.c
aci.c       at.c         backglue.c  back-ndb      back-sock    component.c   ctxcsn.c   filter.c       limits.c         modrdn.c  oidm.c         referral.c    schema_check.c  shell-backends  slapd.conf    slapschema.c  txn.c
acl.c       attr.c       back-hdb    back-null     back-sql     component.h   daemon.c   filterentry.c  lock.c           mods.c    operational.c  result.c      schema_init.c   slapacl.c       slapd.ldif    slaptest.c    unbind.c
aclparse.c  ava.c        back-ldap   backover.c    bconfig.c    config.c      DB_CONFIG  frontend.c     main.c           module.c  operation.c    root_dse.c    schemaparse.c   slapadd.c       slapdn.c      sl_malloc.c   user.c
ad.c        back-bdb     back-ldif   back-passwd   bind.c       config.h      delete.c   globals.c      Makefile         mra.c     overlays       saslauthz.c   schema_prep.c   slapauth.c      slap.h        starttls.c    value.c
add.c       back-dnssrv  back-mdb    back-perl     cancel.c     connection.c  dn.c       index.c        Makefile.in      mr.c      passwd.c       sasl.c        search.c        slapcat.c       slapi         str2filter.c  zn_malloc.c
alock.c     backend.c    back-meta   back-relay    ch_malloc.c  controls.c    entry.c    init.c         matchedValues.c  nt_svc.c  phonetic.c     schema        sets.c          slapcommon.c    slapindex.c   syncrepl.c

```

**Comment by davidcoutadeur** [comment]
Hello,

Strange, the path seems ok, but the required libraries cannot be found.

Please check while in ppm extracted directory, the following path exist:

||../../../libraries/liblber/.libs| ||../../../libraries/libldap_r/.libs|

||

Also, is OpenLDAP correctly compiled?

David


Le 01/07/2020 à 16:59, Dave a écrit :

**Comment by davama** [comment]
>Please check while in ppm extracted directory, the following path exist:
 ||../../../libraries/liblber/.libs| ||../../../libraries/libldap_r/.libs|

No, they do not exist. Neither does it exist in centos7 when build there. The parent folder `liblber` and `libldap_r` exist on both but not the hidden file `.libs`

I tested the exact same procedures on Centos7 and I was able to install ppm with no issues.

You can see good install output here: https://pastebin.com/0uR6Uezd
Bad install output here: https://pastebin.com/wzQKZTtN

If you `vimdiff` them side by side, you only see a noticeable difference on `make CONFIG...` cmds under ppm folder. (line 744 and 729 respectively)

Hope this is helpful.

Thank you very much for the support!

**Comment by davidcoutadeur** [comment]
Seems strange because we build ppm on CentOS 8 (see Clement's answer and 
spec file).
Maybe this is due to a difference between OpenLDAP community source code 
and symas source code?

Could you test with community source code?

David

**Comment by davama** [comment]
Thank you for the quick reply!

> Could you test with community source code?

I tried it but it failed under centos8. Pass with centos7. Used source files from here: https://ltb-project.org/download

**But i figured it out!!!** :tada: :tada: :tada:


So if i switch from using `symas-openldap-devel` to centos' `openlda-devel` it works!
I also did not need change the syntax for `make CONFIG=...`. Used the same syntax from what is documented in the INSTALL file. 

So bottom line you cannot use `symas-openldap-devel`. Use regular centos provided `openldap-devel` package. 
Whether you use ltb source or symas source, they both build without issues. (*side note*: my actual ldap server is still symas-openldap code)

@coudot and @davidcoutadeur 
Thank you both very much for the kind support!
Stay safe

Dave
